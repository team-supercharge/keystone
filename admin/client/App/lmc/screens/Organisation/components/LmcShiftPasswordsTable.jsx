import React from 'react'
import PropTypes from 'prop-types'
import { BlankState } from '../../../../elemental'

const NO_SHIFTS_MESSAGE = "You haven't added any shift passwords yet"
const SUPPORT_LINK = 'https://support.logmycare.co.uk/the-care-office/finishing-your-essential-setup/how-do-i-set-up-a-shift-password'

const renderTable = (shifts) => {
    return (
        <div>
            <h2 style={styles.title}>
                Shift Passwords
            </h2>
            <div className='lmc-theme-gradient' style={styles.divider} />
            <table cellPadding="0" cellSpacing="0" className="Table ItemList">
                <thead>
                    <tr>
                        <th>
                            Title
                        </th>
                        <th>
                            Start
                        </th>
                        <th>
                            End
                        </th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>
                            Shift 1
                        </td>
                        <td>
                            08:99
                        </td>
                        <td>
                            13:22
                        </td>
                    </tr>
                </tbody>
            </table>
            {/* <ul style={styles.list}>
            { JSON.stringify(shifts) }
            </ul> */}
        </div>
    )
}

const renderEmptyPage = () => {
    return (
        <div>
            <BlankState
                heading={NO_SHIFTS_MESSAGE}
                style={styles.noShiftsMessage}
            >
                For extra security, we highly recommend using shift passwords.
                <a 
                    href={SUPPORT_LINK}
                    style={styles.supportLink}
                    target='_blank'
                >
                    Find out more
                </a> 
            </BlankState>
        </div>
    )
}

const LmcShiftPasswordsTable = ({ shifts }) => {
    const hasShifts = shifts && shifts.length
    
    return ( hasShifts ? renderTable(shifts) : renderEmptyPage() )
}

const styles = {
    title: {
        marginBottom: '0.3em',
        fontWeight: 300,
    },
    divider: {
        height: 2,
        marginBottom: 22,
        width: '100%',
    },
    list: {
        listStyle: 'none',
        listStyleType: 'none',
        padding: 0,
    },
    noShiftsMessage: {
        position: 'relative',
        top: 50,
    },
    supportLink: {
        paddingLeft: 5,
    },
}

LmcShiftPasswordsTable.propTypes = {
    shifts: PropTypes.array
}

export default LmcShiftPasswordsTable