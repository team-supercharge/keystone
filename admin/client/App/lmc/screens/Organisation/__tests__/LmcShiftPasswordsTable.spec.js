import React from 'react'
import { shallow } from 'enzyme'
import LmcShiftPasswordsTable from '../components/LmcShiftPasswordsTable.jsx'

describe('LmcShiftPasswordsTable', () => {
    let wrapper
    let shifts
    let savedKeystone
    const onDelete = jest.fn()

    beforeAll(() => {
        savedKeystone = global.Keystone
        global.Keystone = { adminPath: '/admin' }
    })

    beforeEach(() => {
        shifts = [
            { id: 'testId1', title: 'testShift1', password: 'testPassword1' },
            { id: 'testId2', title: 'testShift2', password: 'testPassword2' }
        ]

        wrapper = shallow(
            <LmcShiftPasswordsTable
                shifts={shifts}
                onDelete={onDelete}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders an unordered list', () => {
        expect(wrapper.find('ul').length).toEqual(1)
    })

    it('renders an LmcShiftPasswordItem for each shift', () => {
        expect(wrapper.find('LmcShiftPasswordItem').length).toEqual(
            shifts.length
        )
    })

    it('shows a message + link to support article for no shifts', () => {
        const emptyWrapper = shallow(
            <LmcShiftPasswordsTable
                shifts={undefined}
                onDelete={onDelete}
            />
        )

        const blankTile = emptyWrapper.find('BlankState')
        expect(blankTile.props().heading).toEqual("You haven't added any shift passwords yet")

        const link = emptyWrapper.find('a')
        const supportLink = 'https://support.logmycare.co.uk/the-care-office/finishing-your-essential-setup/how-do-i-set-up-a-shift-password'
        expect(link.props().href).toEqual(supportLink)
    })

    afterAll(() => {
        global.Keystone = savedKeystone
    })
})