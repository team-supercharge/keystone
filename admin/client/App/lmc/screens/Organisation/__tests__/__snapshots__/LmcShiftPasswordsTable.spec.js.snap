// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcShiftPasswordsTable renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2
      style={
        Object {
          "fontWeight": 300,
          "marginBottom": "0.3em",
        }
      }
    >
      Shift Passwords
    </h2>
    <div
      className="lmc-theme-gradient"
      style={
        Object {
          "height": 2,
          "marginBottom": 22,
          "width": "100%",
        }
      }
    />
    <table
      cellPadding="0"
      cellSpacing="0"
      className="Table ItemList"
    >
      <thead>
        <tr>
          <th>
            Title
          </th>
          <th>
            Start
          </th>
          <th>
            End
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Shift 1
          </td>
          <td>
            08:99
          </td>
          <td>
            13:22
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "nodes": Array [
    <div>
      <h2
        style={
          Object {
            "fontWeight": 300,
            "marginBottom": "0.3em",
          }
        }
      >
        Shift Passwords
      </h2>
      <div
        className="lmc-theme-gradient"
        style={
          Object {
            "height": 2,
            "marginBottom": 22,
            "width": "100%",
          }
        }
      />
      <table
        cellPadding="0"
        cellSpacing="0"
        className="Table ItemList"
      >
        <thead>
          <tr>
            <th>
              Title
            </th>
            <th>
              Start
            </th>
            <th>
              End
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Shift 1
            </td>
            <td>
              08:99
            </td>
            <td>
              13:22
            </td>
          </tr>
        </tbody>
      </table>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <LmcShiftPasswordsTable
        shifts={
          Array [
            Object {
              "id": "testId1",
              "password": "testPassword1",
              "title": "testShift1",
            },
            Object {
              "id": "testId2",
              "password": "testPassword2",
              "title": "testShift2",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "shifts": Array [
            Object {
              "id": "testId1",
              "password": "testPassword1",
              "title": "testShift1",
            },
            Object {
              "id": "testId2",
              "password": "testPassword2",
              "title": "testShift2",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2
            style={
              Object {
                "fontWeight": 300,
                "marginBottom": "0.3em",
              }
            }
          >
            Shift Passwords
          </h2>
          <div
            className="lmc-theme-gradient"
            style={
              Object {
                "height": 2,
                "marginBottom": 22,
                "width": "100%",
              }
            }
          />
          <table
            cellPadding="0"
            cellSpacing="0"
            className="Table ItemList"
          >
            <thead>
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Start
                </th>
                <th>
                  End
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  Shift 1
                </td>
                <td>
                  08:99
                </td>
                <td>
                  13:22
                </td>
              </tr>
            </tbody>
          </table>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2
            style={
              Object {
                "fontWeight": 300,
                "marginBottom": "0.3em",
              }
            }
          >
            Shift Passwords
          </h2>
          <div
            className="lmc-theme-gradient"
            style={
              Object {
                "height": 2,
                "marginBottom": 22,
                "width": "100%",
              }
            }
          />
          <table
            cellPadding="0"
            cellSpacing="0"
            className="Table ItemList"
          >
            <thead>
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Start
                </th>
                <th>
                  End
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  Shift 1
                </td>
                <td>
                  08:99
                </td>
                <td>
                  13:22
                </td>
              </tr>
            </tbody>
          </table>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcShiftPasswordsTable
    shifts={
      Array [
        Object {
          "id": "testId1",
          "password": "testPassword1",
          "title": "testShift1",
        },
        Object {
          "id": "testId2",
          "password": "testPassword2",
          "title": "testShift2",
        },
      ]
    }
  />,
}
`;
